### PR TITLE
Adds a magic monkey to the wizard's den

### PR DESCRIPTION
**Pull request by Contrabang**
## What Does This PR Do

Adds a magic monkey with a random wizard den, replacing the normal monkey that is there. There is a 1/1000 chance for that monkey to be MERLIN THE MAGNIFICENT, which is a very large monkey. Otherwise, they will have a random wizard name, but usually shorter.

What makes him magical you might ask? 
Well of course, they teleport whenever someone tries to hit them! They also happen to be resistant to atmos so that if they teleport outside the wizard den when attacked, they'll be fine.
If they do teleport outside, they'll try to teleport back inside in 5 minutes, no one wants to be out in that lava forever.
Wizards cannot mindswap into them.
If a cling gets ahold of them however, I aint gonna stop them, props to them for getting to the wizard's den somehow...
However, they won't completely invulnerable, if you throw shit at them, or just shoot guns at them, they don't stand a chance. They just doesn't wanna get melee'd.
He also has his own blink spell, if an admin inhabits the body.

Also, monkeys will no longer walk into lava nor chasms.

## Why It's Good For The Game
I love silly things, and this is a funny little feature

## Images of changes

Behold, MERLIN, just after gibbing a skrell.
image

## Testing

I tried to hit him a bunch as both a simple mob, and a human, he teleported away. I watched him try to teleport back into the wizard den a few times until he succeeded

## Changelog
:cl:
add: A magical monkey to the wizard's den.
tweak: Monkies will no longer willingly walk into lava nor chasms.
/:cl:

### CODE REVIEW

**Comment by Coolrune206** [comment]
The monkey in the Wizard Den is for Wizards to test spells on. While this is a nifty concept, it also removes the human practice dummy from the den, as I don't think most wizards are going to be testing their spells on a named Wizard Monkey.
Additionally, instant-kill on attempted mindswap is a massive noob-trap and will result in a few instant round-overs on Wizard rounds. Not fun for the bald Wizard, not fun for the players in the round.

**Comment by Contrabang** [comment]
> The monkey in the Wizard Den is for Wizards to test spells on. While this is a nifty concept, it also removes the human practice dummy from the den, as I don't think most wizards are going to be testing their spells on a named Wizard Monkey. 

Fair. You also can't use a good amount of the spells in the wizard den though, I dont think I've ever actually see someone hit the monkey.

> Additionally, instant-kill on attempted mindswap is a massive noob-trap and will result in a few instant round-overs on Wizard rounds. Not fun for the bald Wizard, not fun for the players in the round.

This can happen only if the monkey has a mind, and the wizard tries to use mindswap on them. The monkey can only have a mind if an admin has given them a mind. If they try to mindswap with an admin, I don't think thats too bad of a punishment.
(Plus I don't think bald wizards are going to try to go for mindswap. But this gibbing part can always be removed and replaced with some more indepth information of why they cant mindswap with them.)

**Comment by Coolrune206** [comment]
> > The monkey in the Wizard Den is for Wizards to test spells on. While this is a nifty concept, it also removes the human practice dummy from the den, as I don't think most wizards are going to be testing their spells on a named Wizard Monkey.
> 
> Fair. You also can't use a good amount of the spells in the wizard den though, I dont think I've ever actually see someone hit the monkey.
> 
> > Additionally, instant-kill on attempted mindswap is a massive noob-trap and will result in a few instant round-overs on Wizard rounds. Not fun for the bald Wizard, not fun for the players in the round.
> 
> This can happen only if the monkey has a mind, and the wizard tries to use mindswap on them. The monkey can only have a mind if an admin has given them a mind. If they try to mindswap with an admin, I don't think thats too bad of a punishment. (I'll remove it anyways)

Admins do have a penchant for possessing the monkey on Wiz rounds to either wish them well or give them a little bit of guidance if they're newer 😜

**Comment by Exavere** [comment]
Monke the Archwizard.

**Review by DGamerL — APPROVED** [review]
Monkey buffs

File: code/modules/mob/living/carbon/human/npcs.dm [hunk 1]
```diff
@@ -21,3 +21,90 @@
 	real_name = name
 	resize = 0.8
 	update_transform()
+
+/mob/living/carbon/human/monkey/magic
+	var/return_timer
```
Comment (lewcc) [inline code/modules/mob/living/carbon/human/npcs.dm:26]: autodoc this

File: code/modules/mob/living/carbon/human/npcs.dm [hunk 2]
```diff
@@ -21,3 +21,90 @@
...
+		turfs += T
+
+	var/turf/picked = pick(turfs)
+	if(!picked || !isturf(picked))
+		return
+
+	visible_message("<span class='warning'>[src] blinks away!</span>", "<span class='danger'>Your instincts kick in, and you blink away!</span>")
+	INVOKE_ASYNC(src, PROC_REF(after_the_attack), picked)
+
+	playsound(get_turf(src), 'sound/magic/blink.ogg', 50, 1)
```
Comment (lewcc) [inline code/modules/mob/living/carbon/human/npcs.dm:84]:
```suggestion
	playsound(get_turf(src), 'sound/magic/blink.ogg', 50, TRUE)
```

File: code/modules/mob/living/carbon/human/npcs.dm [hunk 3]
```diff
@@ -21,3 +21,90 @@
...
+
+	// A monkey, with a blink spell? Preposterous
+	playsound(get_turf(src), 'sound/magic/blink.ogg', 50, 1)
+
+	var/list/turfs = list()
+	var/list/target_turfs = range(src, 5) - range(src, 1)
+	for(var/turf/T in target_turfs)
+		if(isspaceturf(T))
+			continue
+		if(T.density)
```
Comment (lewcc) [inline code/modules/mob/living/carbon/human/npcs.dm:67]:
```suggestion
		if(is_blocked_turf(T))
```

File: code/modules/mob/living/carbon/human/npcs.dm [hunk 4]
```diff
@@ -21,3 +21,91 @@
...
+		return // I'm afraid you do not get a free blink
+	if(blink_away())
+		return COMPONENT_CANCEL_ATTACK_CHAIN // prevent people from attacking the monkey
+
+/mob/living/carbon/human/monkey/magic/proc/blink_away()
+	if(HAS_TRAIT(src, TRAIT_HANDS_BLOCKED))
+		return
+
+	// A monkey, with a blink spell? Preposterous
+	playsound(get_turf(src), 'sound/magic/blink.ogg', 50, 1)
```
Comment (Henri215) [inline code/modules/mob/living/carbon/human/npcs.dm:61]:
```suggestion
	playsound(get_turf(src), 'sound/magic/blink.ogg', 50, TRUE)
```

**Review by S34NW — APPROVED** [review]

**Pull request merged**